File: app/src/main/java/io/github/husseinfo/countin/data/CountModel.java
package io.github.husseinfo.countin.data;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

@Entity(tableName = "count")
public class CountModel {

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    public int id;

    @ColumnInfo(name = "title")
    public String title;

    @ColumnInfo(name = "date")
    public String date;

    public CountModel(String title, String date) {
        this.title = title;
        this.date = date;
    }

    public String getYears() {
        return "1";
    }

    public String getMonths() {
        return "2";
    }

    public String getDays() {
        return "3";
    }

    @NonNull
    @Override
    public String toString() {
        return this.date + ';' + this.title;
    }
}

File: app/src/main/java/io/github/husseinfo/countin/data/CountModel.kt
package io.github.husseinfo.countin.data

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import io.github.husseinfo.countin.activities.format
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.Period
import java.time.ZoneOffset
import java.util.Calendar
import java.util.Date

@Entity(tableName = "count")
class CountModel(
    @field:ColumnInfo(name = "title") var title: String,
    @field:ColumnInfo(name = "date") var date: Long,
    @field:ColumnInfo(name = "with_time") val withTime: Boolean,
    @field:ColumnInfo(name = "icon") val icon: String?,
    @field:ColumnInfo(name = "list") val list: String?,
) {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    var id = 0

    fun dateDiff(): Period {
        return Period.between(
            LocalDateTime.ofInstant(
                Instant.ofEpochMilli(date),
                ZoneOffset.systemDefault()
            ).toLocalDate(),
            LocalDate.now(ZoneOffset.systemDefault())
        )
    }

    private fun milliDiff(): Long {
        return Calendar.getInstance().time.time - date
    }

    fun formatDate(): CharSequence {
        val c = Calendar.getInstance()
        c.time = Date.from(Instant.ofEpochMilli(date))
        return c.format(withTime)
    }

    fun getPeriod(): String {
        val difference: Period = dateDiff();
        val per = "${difference.years}Y, ${difference.months}M, ${difference.days}D"
        if (withTime) {
            val diff = milliDiff() / 1000
            val h = (diff / 3600 % 24).toString()
            val m = (diff / 60 % 60).toString()
            return "$per  -  ${h}H, ${m}M"
        }
        return per
    }
}
